Add temporary directory cleanup workarounds

droberts195 opened the pull request:
On some Linux distributions tmpfiles.d cleans files and
directories under /tmp if they haven't been accessed for
10 days.

This can cause problems for ML as ML is currently the only
component that uses the temp directory more than a few
seconds after startup. If you didn't open an ML job for
10 days and then tried to open one then the temp directory
would have been deleted.

This commit prevents the problem occurring in the case of
Elasticsearch being managed by systemd, as systemd private
temp directories are not subject to periodic cleanup (by
default).

Additionally there are now some docs to warn people about
the risk and suggest a manual mitigation for .tar.gz users.

Fixes #31732

droberts195 committed:
Add temporary directory cleanup workarounds

On some Linux distributions tmpfiles.d cleans files and
directories under /tmp if they haven't been accessed for
10 days.

This can cause problems for ML as ML is currently the only
component that uses the temp directory more than a few
seconds after startup. If you didn't open an ML job for
10 days and then tried to open one then the temp directory
would have been deleted.

This commit prevents the problem occurring in the case of
Elasticsearch being managed by systemd, as systemd private
temp directories are not subject to periodic cleanup (by
default).

Additionally there are now some docs to warn people about
the risk and suggest a manual mitigation for .tar.gz users.

elasticmachine commented:
Pinging @elastic/es-core-infra

jasontedor commented on docs/reference/setup/important-settings/es-tmpdir.asciidoc:
```diff
@@ -0,0 +1,23 @@
+However, if you intend to run the `.tar.gz` distribution on Linux for an
+extended period then you should consider creating a dedicated temporary
+directory for Elasticsearch that is not under a path that will have old files
+and directories cleaned from it.  This directory should have permissions set
```
An incredibly minor nit, but for consistency: reduce these two spaces to one.

jasontedor reviewed (approved):
LGTM. I left a nit, no need for another round.

jasontedor commented:
@elasticmachine test this please

droberts195 committed:
Replace 2 spaces with 1

droberts195 commented:
Looks like it's not as simple as I thought.  Elasticsearch itself happily started up under systemd with the extra setting when I tested this on a cloud server, but the packaging tests are consistently failing as some other part of the Elasticsearch script ecosystem isn't being redirected to the private directory.

droberts195 committed:
Don't create packaging config test directories under /tmp

When Elasticsearch runs under systemd the test scripts have a
different view of /tmp to Elasticsearch itself

droberts195 commented:
@elasticmachine run gradle build tests

droberts195 committed:
Merge branch 'master' into tmpdir_removal_problem

droberts195 commented:
@elasticmachine run all packaging tests
